### ISSUE TITLE
Question: launching codeserver with a hierarchy of modules

### ISSUE THREAD

**Issue by dnouls** [tbroyer/gwt-maven-plugin#109]
I am using this plugin on a large project, where I have many gwt-modules not in the same level.
When launching codeserver using -pl myapp-app -am it does not see changes that are done on gwt libs modules that are declared at a different level.

small overview of my modules:
```
project
  common
    widgets-gwt (gwt-lib)
  admin
     admin-backend
     admin-shared
     admin-war
     admin-gwt (gwt-lib)
     admin-app (gwt-app)
```

admin-app uses admin-gwt and widget-gwt

When I launch codeserver I do this from the admin module using:
mvn gwt:codeserver -pl admin-app -am
when I make changes in admin-gwt and reload the browser then I see the codeserver compiling the changes.

But when I changes to widget-gwt, I need to stop the codeserver, rebuild the widget-gwt project and restart the codeserver before I see these changes.

How can I avoid this ? It is quite time confusing and error prone.
Is the only solution to move all the gwt libs in the same parent module as the gwt-app ?

**Comment by tbroyer** [comment]
Could it be #31? (maybe https://github.com/tbroyer/gwt-maven-plugin/issues/107#issuecomment-363460344 would be easier to follow)

**Comment by dnouls** [comment]
I read through those 2 bug reports, but I don’t see how it solves my problem.
All my gwt-libs are using packaging gwt-lib and I am declaring all the dependencies with type gwt-lib in my gwt-app.
I also tried to use classpathScope set to compile+runtime but it does not make any difference.
The only gwt-lib changes that are detected by the codeserver are the ones that are next to the app module.

**Comment by tbroyer** [comment]
I'm assuming your run Maven from the top-level of the project? If you run it from the `admin` subfolder, then this is the expected behavior (that's just how Maven works). Was `widgets-gwt` built as part of the reactor? (you'll see it in the logs if that's the case, otherwise it's just a dependency take from your local repository, so it's loaded as a JAR, and again what you're experiencing would be the expected behavior)

BTW, as a rule of thumb, do **never ever** run `mvn install` unless you fully understand the implications (and always run Maven commands from the root of a multi-module project).

**Comment by dnouls** [comment]
Ok, I see. I am indeed not starting the codeserver from the toplevel of the project but from the admin folder.
How do I configure the top level pom and how do I invoke the codeserver when I am using nested maven modules ?

I need to start the codeserver with admin/admin-app but maven does not seem to allow this. In reality the nesting is even deeper.

What are the implications when we do use mvn install ? most people use it instead of verify. Also since we are developing multiple depending maven projects at the same time, it is almost impossible not to use mvn install or we don’t see the changes across these different projects.

**Comment by tbroyer** [comment]
> Ok, I see. I am indeed not starting the codeserver from the toplevel of the project but from the admin folder.
> How do I configure the top level pom and how do I invoke the codeserver when I am using nested maven modules ?
> 
> I need to start the codeserver with admin/admin-app but maven does not seem to allow this. In reality the nesting is even deeper.

The value of `<projects>` is a comma-delimited list of project IDs, similar to the ones you can give to `mvn -pl` (except that one also accepts relative paths to POMs or their containing directories): either bare artifact IDs (possibly prefixed with ":"), or `${groupId}:${artifactId}`.
AFAICT, nesting shouldn't matter.

> What are the implications when we do use mvn install ? most people use it instead of verify. Also since we are developing multiple depending maven projects at the same time, it is almost impossible not to use mvn install or we don’t see the changes across these different projects.

This is a good read: http://blog.lexspoon.org/2012/12/recursive-maven-considered-harmful.html
tl;dr: as soon as you `mvn install` something (a `-SNAPSHOT`), Maven will later find it. It'll likely be outdated, or built from another branch, etc. Given Maven's design, and strange implementation choices, it might even prefer an artifact from the local repository to a project from the reactor build (this of course also depends on the version of Maven you're using). This is particularly true for `gwt:codeserver` (and `jetty:run` et al.) that forks a build that doesn't reach the `package` phase (and as such doesn't attach an artifact to the project; but we specifically don't want packaged artifacts here, and we don't want to run tests either)

Which leads me to my rules of thumb when working with multi-module projects:
 * never ever install a SNAPSHOT; corollary: never ever deploy a SNAPSHOT, unless you specifically need to share it with other projects.
 * always work from the root of your multi-project build, making good use of `-pl` and `-am` (and `-amd`) and, unfortunately, your IDE (want to run tests for one of your submodules only? good luck doing that efficiently with Maven, when with any other build tool you'd just say "run tests for that submodule" and it'll build the necessary dependencies before that)

(the last rule is “if you have the opportunity to use something else that Maven, don't hesitate”, but I don't want to sound too ranty :wink:)

Rules are meant to be broken though, but understand the implications **before** you do so (using `mvn install` to test the effects of a change in a library into another project that makes use of it is OK, but make sure you cleanup after yourself; and you could also use an adhoc reactor build just for that instead of installing the library)

**Comment by dnouls** [comment]
Thanks for all this info. I am not allowed to move to gradle at this point. But maybe I can experiment a bit to show the benefits to my team. Building a large project top level in maven is most of the time not very incremental when you start using plugins.

Kind regards,
David Nouls
Nouls Consulting BVBA

**Comment by natros** [comment]
Do you know or recommend a gwt plugin for gradle that is as good as yours?

**Comment by dnouls** [comment]
I implemented a toplevel pom file and am now launching the codeserver from there.
I am also using -pl and -am to build the war and launch the codeserver and that seems to work efficiently!
Thanks for the help
